### ISSUE TITLE
Rust-analyzer in VSCode seems to initiate the LSP check too soon when using a devcontainer. 

### ISSUE THREAD

**Issue by dlouwers** [rust-lang/rust-analyzer#16837]
Hi,

Using a devcontainer with nix to get a stable environment to do rust development on. As a `postCreateCommand` I am using `nix-env -i rustup && rustup default stable` (yes, not entirely deterministic but that's not what it is about). This is the first hook after the development container is run. The next being `postStartCommand` which should run after the container has started and `postAttachCommand` which should run after VSCode has attached to the container and has a shell. However the rust-analyzer plugin tries to run `cargo` even before the container has started. Not sure if this is something that can be or you would like to address. Perhaps you'd advise me to approach this differently.

Kind regards,

Dirk Louwers

**Comment by dlouwers** [comment]
Note for anyone that runs into this: Apparently the `updateContentCommand` does work for this even though one would think any command before `start` would do. Not sure if this is intentional so it could just be a workaround.

**Comment by dlouwers** [comment]
I git a bit further with this. It seems like the VSCode plugin does not use the default shell when executing commands. This causes the rust toolchain not to be available on the path. In my case I use a nix-shell which has been configured for VSCode. Is there a setting to change the shell that rust-analyzer uses?

**Comment by dlouwers** [comment]
Thanks. The challenge is that the environment itself isn't available until nix-shell reads the nix file and pulls in dependencies. So I can not hardcode a path. I can override all commands if neccesary though. For example `nix-shell --command 'cargo check --quiet --workspace --message-format=json --all-targets'` works fine.

**Comment by dlouwers** [comment]
Tried providing a full path to `nix-shell` too in case it really runs outside of a shell. No luck. It seems to completely ignore the `overrideCommand` setting.

**Comment by dlouwers** [comment]
I am using nix as a container feature as recommended. That grants composability but also means I don't get to intervene in its installation (see: https://github.com/devcontainers/features/tree/main/src/nix).

**Comment by dlouwers** [comment]
Would be nice to figure out how to (preferably) override the shell that is used by rust-analyzer plugin or override how it runs its commands.

**Comment by lnicola** [comment]
There's no shell. You can spawn processes without one.

**Comment by lnicola** [comment]
The question isn't how to make rust-analyzer use a shell when it doesn't need one, but rather how to make `nix-shell` dump its environment somewhere the Code server can find. Maybe even `nix-shell --command env > /etc/profile.d/devcontainer.sh`.

**Comment by dlouwers** [comment]
rust-analyzer doesn't source `etc/profile` and by extension `etc/profile.d/devcontainer.sh`. If, as you say, no shell is involved, there is also no environment and nothing gets sourced. After doing some research it turns out `/etc/profile` is pretty much associated with the bash shell.  (See: https://manpages.org/bash#files)
> During an interactive shell login, if .bash_profile is not present in the home directory, Bash looks for .bash_login. If found, Bash executes it. If .bash_login is not present in the home directory, Bash looks for .profile and executes it.

Can you tell me what process(es) rust-analyzer is trying to start as what user? Or point me to the area in the source where this happens. That would allow me to solve my issue without having to do too much reverse engineering/probing.

Thanks in advance.

**Comment by lnicola** [comment]
> rust-analyzer doesn't source etc/profile and by extension etc/profile.d/devcontainer.sh

It doesn't, but I assumed Code or someone does, otherwise how would the steps in that article work? But see below.

> After doing some research it turns out /etc/profile is pretty much associated with the bash shell.

Yes, mostly. There's also [`userEnvProbe`](https://containers.dev/implementors/json_reference/), which might read it, but I don't really understand what it does. The article above does set it, though.

> Can you tell me what process(es) rust-analyzer is trying to start

`cargo`, `rustc` and `rustup`, around https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/src/toolchain.ts#L169. and `rust-analyzer` itself via https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/src/bootstrap.ts#L34. Note the `patchelf` hack for Nix.

**Comment by dlouwers** [comment]
> It doesn't, but I assumed Code or someone does, otherwise how would the steps in that article work?
The article is very dated. A lot of things in there no longer work and or have been deprecated.

> Yes, mostly. There's also [userEnvProbe](https://containers.dev/implementors/json_reference/), which might read it, but I don't really understand what it does. The article above does set it, though.
userEnvProbe determines what type of shell should be started which in turn determines what gets sourced. Strangely my setting is `loginInteractiveShell` and I would expect `/etc/profile` to therefore get sourced but ofc I have no idea if VSCode respects this. It only sources `~/.bashrc` when I start a terminal.

>cargo, rustc and rustup, around https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/src/toolchain.ts#L169. and rust-analyzer itself via https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/src/bootstrap.ts#L34. Note the patchelf hack for Nix.
Luckily I am not running NixOS, only Nix package manager on Ubuntu so that code gets skipped. Looking at `toolchain.ts` it is indeed running in a shell context because it is able to read environment variables. But I have added the cargo pod to about every global settings file imaginable: /etc/profile, /etc/bashrc, /etc/bash.bashrc, none seem to get sourced. I wish there was a way to trick rust-analyzer plugin to echo `$(whoami)` and `$SHELL` so I could figure out what context to pin down. A quick search for the shell and user being used by VSCode extensions didn't yield any immediate result. Most documentation is monkey-see-monkey-do rather than in depth documentation on what's really happening. Perhaps I should just write a VSCode extension to answer those question empirically myself if there is no documentation on it.

> The current (Code) user, of course. We're not in the business of making system changes.
I am asking because in a devcontainer context the user is NOT the local VSCode user. Default for VSCode is the `vscode` user with `sudo` privileges.

**Comment by lnicola** [comment]
> It only sources ~/.bashrc when I start a terminal.

I see. Another approach might be https://code.visualstudio.com/remote/advancedcontainers/environment-variables#_option-2-use-an-env-file: you could run `nix-shelll --run-command env` save that to a file, then set the output as the `env_file`. Didn't try it, of course.

> Looking at toolchain.ts it is indeed running in a shell context because it is able to read environment variables.

No, it doesn't. Environment variables exist outside the shell. We even mention this in the [user manual](https://rust-analyzer.github.io/manual.html#rust-analyzer-language-server-binary):

> If your editor can’t find the binary even though the binary is on your $PATH, the likely explanation is that it doesn’t see the same $PATH as the shell, see [this issue](https://github.com/rust-lang/rust-analyzer/issues/1811). On Unix, running the editor from a shell or changing the .desktop file to set the environment should help.

My preferred method under a DE is to use `~/.config/environment.d`.

> I wish there was a way to trick rust-analyzer plugin to echo $(whoami) and $SHELL so I could figure out what context to pin down.

I'd expect something like `ps aux | grep rust-analyzer` and `cat /proc/$(pidof rust-analyzer)/environ | tr '\0' $'\n'` (in `bash` syntax) to answer that without a new extension.

> I am asking because in a devcontainer context the user is NOT the local VSCode user.

It runs under the remote user, the one that's running the Code server. But it never calls `sudo`, nor `setuid()`.

**Comment by dlouwers** [comment]
> I see. Another approach might be https://code.visualstudio.com/remote/advancedcontainers/environment-variables#_option-2-use-an-env-file: you could run nix-shelll --run-command env save that to a file, then set the output as the env_file. Didn't try it, of course.

Prefer a solution that isn't within the repo.

> No, it doesn't. Environment variables exist outside the shell. We even mention this in the [user manual](https://rust-analyzer.github.io/manual.html#rust-analyzer-language-server-binary):

Thanks, didn't know that. Will try and look into how those type of environments are set up.

> My preferred method under a DE is to use ~/.config/environment.d

Sounds promising. Do all key value pairs in files in this directory get loaded in the context the extension tries to load binaries? Will give it a try by adding a test key/value pair to a test.env file there and see where it becomes available.

**Comment by dlouwers** [comment]
Tried:
- ~/.config/environment.d 
- added ~/.config/environment that sources all files in ~/.config/environment.d
- added to /etc/environment
None of these variables became available.

Tried the `runArgs` route but that doesn't work since the paths aren't known when the container is started so the devcontainer.env only gets populated after start which is too late to use it as a run argument.

Running out of ideas here.

**Comment by lnicola** [comment]
> Prefer a solution that isn't within the repo.

Does it have to? You could have a shell script that runs `nix-shell --command env` in your container and produces a `.env` file, which you don't check in to the repo.

> Sounds promising. Do all key value pairs in files in this directory get loaded in the context the extension tries to load binaries?

No, that's only a DE thing. GNOME and KDE will load them. SSH and console login won't.

**Comment by dlouwers** [comment]
> Does it have to? You could have a shell script that runs nix-shell --command env in your container and produces a .env file, which you don't check in to the repo.

If this is .env is passed to `runArgs` it isn't present yet at the time the container starts so it fails like I stated in my last comment.
I determine the `cargo` path with `$(nix-shell shell.nix --run "dirname \$(which cargo)")` and that can only be known after container startup which is after `runArgs` are passed.

**Comment by lnicola** [comment]
> If this is .env is passed to runArgs it isn't present yet at the time the container starts so it fails like I stated in my last comment.

Yeah, I meant creating it before using the container, but without checking it in.

**Comment by dlouwers** [comment]
> Yeah, I meant creating it before using the container, but without checking it in.

Yeah I'd love to but `shell.nix` is mounted into the container so I can't.

**Comment by dlouwers** [comment]
Will see if I can get more info from the authors (https://hub.docker.com/_/microsoft-devcontainers-base).

Meanwhile I am thinking of trying a workaround of actually symlinking rust binaries to places that commonly already are on the PATH since I can not seem to be able to influence the environment at all.
